### PR TITLE
Introduce `hir::ExprKind::Let` - Take 2

### PR DESCRIPTION
**Pull request by c410-f3r**
Builds on #68577 and depends on #79328.

cc #53667

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/expr/as_rvalue.rs
Comment (matthewjasper) [inline compiler/rustc_mir_build/src/build/expr/as_rvalue.rs:59]: This can be moved to the branch at the end

File: compiler/rustc_ast_lowering/src/expr.rs
Comment (matthewjasper) [inline compiler/rustc_ast_lowering/src/expr.rs:91]: If expressions will need to wrap anything that's not a `Let` expression in `DropTemps` and `compiler/rustc_passes/src/region.rs` will need updating to make the condition not be a terminating scope.

**Comment by c410-f3r** [comment]
Hi @matthewjasper. Thanks for the review!

This current status of this PR is "pretty bare bones". Sorry for the lack of work :P
Tomorrow I will address the comments and start coding again.

File: src/test/ui/pattern/usefulness/deny-irrefutable-let-patterns.stderr
Comment (petrochenkov) [inline src/test/ui/pattern/usefulness/deny-irrefutable-let-patterns.stderr:5]: The "irrefutable pattern" lint changed its span from the whole if-let/while-let expression to span of the specific pattern which is irrefutable.
Looks like an acceptable and maybe desirable change.

The "macro" tests (`if-let.rs` and `while-let.rs`) also changed for exactly the same reason.

File: src/test/ui/associated-type-bounds/traits-assoc-type-macros.rs
Comment (c410-f3r) [reply src/test/ui/associated-type-bounds/traits-assoc-type-macros.rs:2]: Hehehehe :P
A temporary hack to make the test suite "green". Looks like it is necessary to clean local builds.

File: src/test/ui/pattern/usefulness/deny-irrefutable-let-patterns.stderr
Comment (c410-f3r) [reply src/test/ui/pattern/usefulness/deny-irrefutable-let-patterns.stderr:5]: Maybe add the whole `pattern = expression` span for `hir::ExprKind::Let`?

```
Let {
  expr: Expr,
  pat: Pat,
  span: Span
}
```

Comment (petrochenkov) [reply src/test/ui/pattern/usefulness/deny-irrefutable-let-patterns.stderr:5]: Maybe, if you really want to preserve the old behavior.

Comment (jplatte) [reply src/test/ui/pattern/usefulness/deny-irrefutable-let-patterns.stderr:5]: My 2c: New, small span is better.

File: src/test/ui/expr/if/if-let.stderr
Comment (c410-f3r) [inline src/test/ui/expr/if/if-let.stderr:5]: Span comes from AST and for some reason, macro tokens definitions aren't being collected . Normal `let pat = expr` span don't crop.

File: src/test/ui/expr/if/if-let.stderr
Comment (c410-f3r) [inline src/test/ui/expr/if/if-let.stderr:56]: Expected. Match expressions have the whole desugared `match scrutinee { ... }` span but `Let(..)` does not.

File: src/test/ui/expr/if/if-ret.stderr
Comment (c410-f3r) [inline src/test/ui/expr/if/if-ret.stderr:5]: The path take takes care of the lint is now generic over `if` and `while`.

File: src/test/ui/rfc-2294-if-let-guard/feature-gate.stderr
Comment (c410-f3r) [inline src/test/ui/rfc-2294-if-let-guard/feature-gate.stderr:196]: Centril also ignored theses tests https://github.com/rust-lang/rust/pull/68577/files#diff-95f7fe3af3dc85b9699c156fbdd6afacadeb79bc450a1890481110d621d15d01. Not sure if they should be restored.

File: src/test/ui/rfc-2497-if-let-chains/protect-precedences.stderr
Comment (c410-f3r) [inline src/test/ui/rfc-2497-if-let-chains/protect-precedences.stderr:19]: This addition is a good thing IMO

File: src/test/ui/pattern/issue-82290.rs
Comment (c410-f3r) [inline src/test/ui/pattern/issue-82290.rs:6]: Originally created because of an ICE -> https://github.com/rust-lang/rust/issues/82290

File: src/test/ui/issues/issue-37576.stderr
Comment (c410-f3r) [inline src/test/ui/issues/issue-37576.stderr:58]: Not sure if this diagnostic is desirable

**Review by c410-f3r — COMMENTED** [review]
Ok, these are the last "details" of the Rustc tests.

@matthewjasper I still need to fix Clippy but it shouldn't take long

File: compiler/rustc_ast/src/ast.rs
Comment (matthewjasper) [inline compiler/rustc_ast/src/ast.rs:1295]: Can you add what the span points to  to the comment here?

**Comment by c410-f3r** [comment]
@matthewjasper Gave all that I could to fix Clippy but couldn't proceed further. Is it possible to merge this PR the way it currently is? I promise to fix all few remaining tests upstream with the help of someone that knows Clippy better.

File: src/tools/clippy/clippy_lints/src/option_if_let_else.rs
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/option_if_let_else.rs:94]: Also here `&expr.kind` over `ref` everywhere. And big NIT: swapping `expr` and `stmts` is unnecessary and makes the code slightly less intuitive (stmts always come before exprs).

File: src/tools/clippy/clippy_lints/src/ranges.rs
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/ranges.rs:324]: I strongly prefer to keep pattern matching those. I don't see a reason why using indexing here should have any advantage.

File: src/tools/clippy/clippy_lints/src/returns.rs
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/returns.rs:215]: Why can we just stop handling this case?

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/higher.rs:435]: No newline at end of file.

File: src/tools/clippy/clippy_utils/src/hir_utils.rs
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/hir_utils.rs:696]: Why not also hash the pattern?

File: src/tools/clippy/clippy_utils/src/hir_utils.rs
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/hir_utils.rs:711]: Why can this be removed? And if it can, please remove it completely instead of commenting it out.

File: src/tools/clippy/clippy_utils/src/lib.rs
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/lib.rs:872]: Why can we stop handling `if let` here?

File: src/tools/clippy/tests/ui/match_overlapping_arm.stderr
Comment (flip1995) [inline src/tools/clippy/tests/ui/match_overlapping_arm.stderr:62]: This must not lint.

**Review by flip1995 — CHANGES_REQUESTED** [review]
Since the Clippy changes are really big, I'd appreciate another pair of eyes from someone from @rust-lang/clippy. I did a in-depth review, but a sanity check of the changes (possibly after my comments were addressed) would be good.

Test changes should be fine, except for the one comment I left on one test. Otherwise some style changes are necessary and I still have a few questions about changes that look suspicious to me.

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/higher.rs:65]:
```suggestion
        } else {
```

File: src/tools/clippy/clippy_utils/src/hir_utils.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_utils/src/hir_utils.rs:696]: Couldn't find an example that hashes patterns? Can you point one?

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_utils/src/higher.rs:65]: Is it possible to run rustfmt while developing Rustc? `./x.py fmt` doesn't touch `src/tools/clippy`

File: src/tools/clippy/clippy_lints/src/loops/never_loop.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_lints/src/loops/never_loop.rs:91]: IIRC, this `ref` stuff was enforced when using all restricted lints.
In any case, it doesn't matter. I will `&expr.kind` for consistence

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/higher.rs:65]: Not 100% sure, but try to `cd` into the clippy dir and then run `cargo dev fmt`.

**Comment by c410-f3r** [comment]
> Since the Clippy changes are really big, I'd appreciate another pair of eyes from someone from @rust-lang/clippy. I did a in-depth review, but a sanity check of the changes (possibly after my comments were addressed) would be good.

Taking aside the commented  test, I will be happy to fix any remaining issue upstream since everything is now passing. In the meanwhile, please consider unblocking the merging of this PR

File: src/tools/clippy/clippy_lints/src/ranges.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_lints/src/ranges.rs:324]: This was carried from previous "hard-to-figure-out" rebases. Fixed

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_utils/src/higher.rs:65]: `cargo fmt` did something but `cargo dev fmt` gives
```
error[internal]: left behind trailing whitespace
     --> ...rust/library/core/src/../../stdarch/crates/core_arch/src/mips/msa.rs:17403:17403:25
```

**Comment by flip1995** [comment]
I'll run this through our lintcheck tool to see if this has any impact on the most popular crates out there. Please also answer the questions above. 

Just merging this without careful consideration may result into a sync mess between Rust and Clippy and will increase the time to get those fixed enormously. IMO merging this should be blocked until `beta` is branched, otherwise we'd ship this directly to stable (in 6 weeks) without having it in nightly for even one week.

File: src/tools/clippy/clippy_utils/src/lib.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_utils/src/lib.rs:872]: `If let` is now a normal if` expr plus `let` plus instead of the previous `match` de-sugaring, so `if` encompasses `if let`

File: src/tools/clippy/tests/ui/match_overlapping_arm.stderr
Comment (c410-f3r) [reply src/tools/clippy/tests/ui/match_overlapping_arm.stderr:62]: They have the same pattern, expression and block return type so why this snippet shouldn't lint? If desirable, I can transfer the code to `if_same_then_else.rs`

File: src/tools/clippy/clippy_lints/src/loops/while_let_loop.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_lints/src/loops/while_let_loop.rs]: I wasn't able to find any resources related to multiple `else if` or `else if let` as this lint doesn't seem to take into account these features. Can you please elaborate?

File: src/tools/clippy/clippy_lints/src/returns.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_lints/src/returns.rs:215]: Same as https://github.com/rust-lang/rust/pull/80357#discussion_r624517410.
Added a `needless_return.rs` test to show that the lint is working as expected -> https://github.com/rust-lang/rust/pull/80357/files#diff-324e8719adb56e60d34436af7504ccdcd23b9acc31d22f8f527810a40f0b8b81

File: src/tools/clippy/clippy_utils/src/hir_utils.rs
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/hir_utils.rs:696]: No, but I think we should also hash `Pat`s, since we also implement `Eq` for it. This is the first `ExprKind` that has a `Pat` field. Could you add a `// FIXME` here, please? No need to implement `Pat` hashing in this PR.

**Comment by c410-f3r** [comment]
> Just merging this without careful consideration may result into a sync mess between Rust and Clippy and will increase the time to get those fixed enormously. IMO merging this should be blocked until beta is branched, otherwise we'd ship this directly to stable (in 6 weeks) without having it in nightly for even one week.

Can you elaborate what could go wrong or cause delays? Every test was already fixed and any possible remaining considerations will be tackled and fixed by me upstream in the following days.

File: src/tools/clippy/tests/ui/match_overlapping_arm.stderr
Comment (flip1995) [reply src/tools/clippy/tests/ui/match_overlapping_arm.stderr:62]: Oh wait, I thought this test was added because it tests that the comments are different. I'm now noticing that this test file shouldn't even test `if_same_then_else`. Can you just add an `#[allow(clippy::if_same_then_else)` to this test file and bless it again, please?

File: src/tools/clippy/tests/ui/needless_return.stderr
Comment (flip1995) [inline src/tools/clippy/tests/ui/needless_return.stderr:87]: Same here: Just allow the lint that should not be tested in this file.

File: src/tools/clippy/clippy_utils/src/hir_utils.rs
Comment (c410-f3r) [reply src/tools/clippy/clippy_utils/src/hir_utils.rs:696]: Ok. Do you mind also opening a PR on `rust-lang/clippy`? FIXMEs or TODOs tend to be forgotten over time.

File: src/tools/clippy/tests/ui/match_overlapping_arm.stderr
Comment (c410-f3r) [reply src/tools/clippy/tests/ui/match_overlapping_arm.stderr]: Done

**Comment by flip1995** [comment]
If this gets merged now, it will get directly into beta. If this should introduce bugs in Clippy that aren't covered by our tests (yet), we'd have not only to sync Rust and Clippy back and forth (which can involve 1h of work in the worst case for each sync), but also have to backport patches you'd make upstream to beta, which is even more involved. All this hours can be avoided by just waiting a few more days. Thursday is the next Rust release, so beta should be branched in a few days AFAIK. If you want I can rebase this for you. Putting time into rebasing this once would save me many hours and headaches that would come from having to backport bug fixes for bugs that may be introduced by this.

It is important to note, that the absence of test failures does not imply the absence of regressions. We had cases before, where changes to rustc introduced regressions in Clippy that took weeks to resolve and weren't covered by Clippy test cases. 

Shipping such a big change in the release week is just a really bad idea. The only reason this is not blocked by CI is that Clippy is a subtree, not a submodule. Otherwise CI wouldn't even allow to merge this PR.

**Review by flip1995 — CHANGES_REQUESTED** [review]
Ok, one more small thing and then I'll run this through lintcheck and the Clippy changes should be good. (I still don't want to get this into beta).

**Comment by c410-f3r** [comment]
@flip1995 Thanks for your detailed explanation. Better play safe than sorry so yeah, let's wait a little longer.
In the meanwhile, feel free to point anything that can enhance the test coverage or avoid possible future problems.

**Comment by flip1995** [comment]
> In the meanwhile, feel free to point anything that can enhance the test coverage or avoid possible future problems.

If I would know where we're missing regression test, we wouldn't miss any regression tests 😄 

---

You'll have to `./x.py test --bless src/tools/clippy` one more time.

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (camsteffen) [inline src/tools/clippy/clippy_utils/src/higher.rs:83]: Why is this called `ast` but it takes a `hir::Expr`?

File: src/tools/clippy/clippy_lints/src/ranges.rs
Comment (camsteffen) [inline src/tools/clippy/clippy_lints/src/ranges.rs:325]: Usually `ref` or `&` is unnecessary with HIR since everything is already a reference. This creates double references.

File: src/tools/clippy/tests/ui/crashes/issues_loop_mut_cond.stderr
Comment (camsteffen) [inline src/tools/clippy/tests/ui/crashes/issues_loop_mut_cond.stderr:7]: Should allow this.

**Review by camsteffen — CHANGES_REQUESTED** [review]
Here's an extra Clippy review. I'm excited to see a move toward let-chains!

Some of the changes seem like unrelated stylistic changes which I think should be avoided in PR's as large as this.

Not Clippy related, but shouldn't `hir::Guard::IfLet` be removed?

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/higher.rs:357]: I wouldn't refactor this in a Rust PR. We can open an issue about it though.

File: src/tools/clippy/clippy_utils/src/higher.rs
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/higher.rs:24]: We could also implement `try_from` for those structs, which would be pretty much the same. I like the `ForLoop::<function>(xyz)` style better. But I would prefer `TryFrom` here. But I would also fix this in Clippy once it's synced back.

**Comment by c410-f3r** [comment]
The requested waiting time has passed, this PR was rebased based on the latest Clippy changes and the majority of the required Clippy code changes were attended without question on my side. Can you guys please unblock the merge of this PR? As previously said, further refinements will be addressed upstream.

**Review by flip1995 — APPROVED** [review]
Clippy changes LGTM now. Everything that is left are just code style things. So from the Clippy side this PR is un-blocked. Thanks a lot for your patience @c410-f3r!  

@matthewjasper Not sure if you want to take another look on the PR, but you can r=me for the Clippy changes. 👍

**Review by matthewjasper — COMMENTED** [review]
r=me with comments addressed and tests updated
@bors rollup=never

**Comment by c410-f3r** [comment]
@matthewjasper Done!

File: compiler/rustc_typeck/src/check/expr.rs
Comment (matthewjasper) [reply compiler/rustc_typeck/src/check/expr.rs:884]: You've changed the parameter in the wrong method here

Comment (c410-f3r) [reply compiler/rustc_typeck/src/check/expr.rs:884]: Sorry, too much coffee with little sleep

**Comment by c410-f3r** [comment]
@matthewjasper Everything should be correctly now

**Comment by c410-f3r** [comment]
```
 error[E0004]: non-exhaustive patterns: `Let { .. }` not covered
   --> compiler/rustc_mir_build/src/thir/visit.rs:28:11
    |
28  |       match expr.kind {
    |             ^^^^^^^^^ pattern `Let { .. }` not covered
```

Not sure why this file wasn't present in the last re-base. Fixed

**Comment by c410-f3r** [comment]
>  error: ran out of registers during register allocation
Spurious error?

Also related -> https://github.com/rust-lang/rust/issues/83854

**Comment by c410-f3r** [comment]
On https://github.com/rust-lang/rust/pull/81942, the same error happened before as stated by https://github.com/rust-lang-ci/rust/runs/2264701480 and should have been fixed by https://github.com/rust-lang/rust/pull/81942/files#diff-dc95efbb37517cceb4b4bb3a4940dabf0add9ab29294cce7865077a484a3003f.

@the8472 @Mark-Simulacrum Any clues?

**Comment by the8472** [comment]
This one is slightly different. The previous case happened in the `#thin0` scenario. This one is failing in `#fat0`. Perhaps some of your changes shuffle things around so that the llvm bug is now triggered by a different case. As your scare quotes probably already acknowledge: we merely papered over the issue, the underlying cause still exists.

The failure was deterministic, so you might have to massage the test parameters until it passes and post an update to #81942. Or exclude the particular platform. It should be reproducible under docker.

Just to keep things up to date it would also be interesting if `#thin0` still fails after your changes. Exposing the bug in additional scenarios seems somewhat concerning.

**Comment by c410-f3r** [comment]
Thank you very much for the detailed explanation @the8472

**Comment by c410-f3r** [comment]
Locally, `fat0` and `thin0` variants show  ```
rustc: /checkout/src/llvm-project/llvm/lib/CodeGen/RegAllocFast.cpp:883: void {anonymous}::RegAllocFast::defineVirtReg(llvm::MachineInstr&, unsigned int, llvm::Register, bool): Assertion `PhysReg != 0 && "Register not assigned"' failed.```, which is similar to what is happening on #81825.

LLVM internals is something totally unknown to me so I won't be able to provide too much help in this field.

**Comment by matthewjasper** [comment]
I think that we should just ignore the test on the failing target: the test is already passing only because of carefully chosen parameters and `-C opt-level=0 -C lto=...` is not a common use case.

**Comment by c410-f3r** [comment]
#81825 was merged. It is time to do the re-base dance again

**Comment by c410-f3r** [comment]
At last! Thanks @matthewjasper

**Comment by rylev** [comment]
This change introduced some [mixed performance results](https://perf.rust-lang.org/compare.html?start=2bd17c1d43bba43412cc2f051323a279d6751e43&end=2a6fb9a4c0e5ca7a81999065943b211c226fe9d8&stat=instructions:u). The largest changes are performance wins, but regressions outweight improvements in terms of numbers. Interestingly, many of the regressions were in full builds, and I'm not quite sure why that would be. 

As part of the performance triage process, I'm marking this as a performance regression. If you believe this performance regression to be justifiable or once you have an issue or PR that addresses this regression, please mark this PR with the label `perf-regression-triaged`.

@rustbot label +perf-regression

**Comment by matthewjasper** [comment]
Hmm, I can't see anything from the by query times, so maybe this is a case where codegen has been slightly regressed resulting in a slower rustc. I'll investigate some more and create a PR soon if I find anything.

**Pull request merged**
